### PR TITLE
fix etcd-io/bbolt#247 move compact logic

### PR DESCRIPTION
**Pull request by andrewpmartinez**
See  #247

### CODE REVIEW

**Comment by andrewpmartinez** [comment]
Build seems to have failed due to a downstream go vet? Not sure how to address it.

**Comment by ptabor** [comment]
@aleksandra-malinowska
@jingyih  

It's interesting. I thought that etcd uses bolt's defragmentation, but it seems it has its own: 
I wonder whether they are fully equivalent, or one of the went out of sync: 
  https://github.com/etcd-io/etcd/blob/b7f0f52a16dbf83f18ca1d803f7892d750366a94/mvcc/backend/backend.go#L349

We should try to align them - e.g. some monitoring hooks might be needed.

**Comment by ptabor** [comment]
> Build seems to have failed due to a downstream go vet? Not sure how to address it.

I think that go 1.12 might be too old for the current state of honnef library: 

https://github.com/etcd-io/bbolt/blob/f6be82302843a215152f5a1daf652c1ee5503f85/.travis.yml#L7

IMHO we should:
  - update go to 1.15 
  - start using https://github.com/myitcv/gobin to poll the tool dependencies.

File: compact.go [hunk 1]
```diff
@@ -0,0 +1,109 @@
+package bbolt
+
+func Compact(dst, src *DB, txMaxSize int64) error {
```
Comment (ptabor) [inline compact.go:3]: Please document contract of a public method.

File: compact.go [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+		// Create buckets on subsequent levels, if necessary.
+		b := tx.Bucket(keys[0])
+		if nk > 1 {
+			for _, k := range keys[1:] {
+				b = b.Bucket(k)
+			}
+		}
+
+		// Fill the entire page for best compaction.
+		b.FillPercent = 1.0
```
Comment (ptabor) [inline compact.go:52]: I wonder whether this line means that the bucket stays forever with the fill factor of 1.0 ? 

E.g. user has default fill factor of pages: 50%... calls compaction, and lands on 100% fill level not only on compacted data (that makes some sense), but also on all future writes.

**Comment by ptabor** [comment]
LGTM - thank you. 
Please squash the commits (.gitignore + go1.15 can go separately).

**Comment by andrewpmartinez** [comment]
> Please squash the commits (.gitignore + go1.15 can go separately).

By go separately, do you mean in a separate squashed commit or in a separate fix all together?

**Comment by ptabor** [comment]
Commit would be sufficient IMHO. Thx.

**Comment by andrewpmartinez** [comment]
Done then.

**Comment by ptabor** [comment]
LGTM - as a recommendation to maintainers (@gyuho, @xiang90, @jingyih @jpbetz)

File: compact.go (hunk 2 above)
Comment (ptabor) [reply compact.go:56]: Optional TODO: To verify/fix that.

File: compact.go [hunk 3]
```diff
@@ -0,0 +1,113 @@
+package bbolt
+
+// Compact will create a copy of the source DB and in the destination DB. This may
+// reclaim space that the source database no longer has use for. txMaxSize can be
+// used to limit the transactions size of this process and may trigger intermittent
+// commits. A value of zero will ignore transaction sizes.
+func Compact(dst, src *DB, txMaxSize int64) error {
```
Comment (ptabor) [inline compact.go:7]: Optional: Please consider adding  TODO: to merge with: 
https://github.com/etcd-io/etcd/blob/b7f0f52a16dbf83f18ca1d803f7892d750366a94/mvcc/backend/backend.go#L349

**Review by ptabor — APPROVED** [review]

File: compact.go (hunk 2 above)
Comment (xiang90) [reply compact.go:56]: for sequential writes, it is optimal to keep the fill factor to 100%. in the compact/copy case, it should be all sequential writes if i remember correctly.

Comment (ptabor) [reply compact.go:56]: Yes. During compaction it make sense to pack the data in the dense way. 
What I'm afraid of is that this settings stays for this bucket forever (post compation). 
IMHO it should go back to what the user had configured pre-compaction.

File: compact.go (hunk 3 above)
Comment (andrewpmartinez) [reply compact.go:8]: Done.

**Comment by andrewpmartinez** [comment]
Do I need to do anything else here?

File: compact.go (hunk 2 above)
Comment (ptabor) [reply compact.go:57]: I miss-assumed the setting is preserved between transactions. But it's scoped to a transaction.

**Comment by ptabor** [comment]
Thank you.

**Pull request merged**